### ISSUE TITLE
OrleansAzureUtils enhancements for table storage

### ISSUE THREAD

**Issue by jason-bragg** [dotnet/orleans#1456]
In preparation for improving the grain state storage in Orleans, I propose the following enhancements to OrleansAzureUtils.

**Partitions and Transactions**
The azure table storage client library provides a set of capabilities that resemble a simplified database.  This provides a model that feels familiar to engineers experienced with database concepts.  Despite this pattern, table storage is not a database, and many of the operations, either through hard constraints or performance constraints, are limited to actions on partitions.  As one of the intentions of our storage layer is to simplify access to table storage, there is an opportunity to simplify the model by reducing the supported operations from actions against rows in partitions in tables on an account to one of actions against rows in a partition.  The below interfaces embody a storage access pattern supporting this approach.

```
public interface ITablePartition
{
    Task<TElement> QueryRow<TElement>(string rowKey)
        where TElement : class, ITableEntity, new();

    Task<IEnumerable<TElement>> QueryRowRange<TElement>(string rowKeyStart, string rowKeyEnd)
        where TElement : class, ITableEntity, new();

    Task Commit(IPartitionTransaction transaction);
}

public enum StorageOperation
{
    Insert,
    Update,
    Delete,
    InsertOrReplace,
}

public interface IPartitionTransaction
{
    IEnumerable<Tuple<StorageOperation, ITableEntity>> Operations { get; }
    Guid ContextId { get; }
}

public interface ITablePartitionManager
{
    Task<ITablePartition> GetPartition(TableName, string partitionKey);
}
```

In the above pattern, a grain or storage provider for a grain, would obtain a table partition at initialization time, then manage entities within that partition by rowkey.  The majority of the application logic would not care about storage account, table, or partition key.  It would need only manage its rows.

This pattern also allows for the introduction, by the partition manager, of caching and monitoring decorators over the ITablePartition interface, without application logic knowing or caring about these behaviors.  It also improves the testability of the grain, as the storage layer can entirely be mocked out.

**Table Entities as business logic objects**
The data types natively supported by table storage entities are fairly limited, and the storage layer is somewhat sensitive to what other data is in an entity.  As business logic often deals with lists, dictionaries and/or other complex and unsupported data types, we need to store this data in a way that will not break the table entity serialization and will allow for these fields to be translated to a format that was of a supported property type.  To do this I suggest the development of an implementation of ITableEntity (OrleansTableEntity?) in which we can better control which properties are stored (via Attributes) and provide our own extensions to the supported property types.

The use of attributes to explicitly mark properties to be persisted provides service developers better control over how table entities are organized and stored.  The introduction of a “Persistent” attribute will allow us to mark properties for storage regardless of their accessibility or the presence of get/set functions.

Example:

```
public class TestEntity : OrleansTableEntity
{
    [Persistent]
    private DateTime FirstTimeUGC; // stored

    [Persistent]
    public DateTime LastTimeUGC;  // stored

    public Guid Id { get; private set; } // not stored
}
```

By implementing our own table entity, we have a means of extending the types of data we can store in tables at the storage layer.  This allows us to support a set of commonly needed, but not natively supported, data types (such as lists, dictionaries, and other containers) to our shared library, as well as allow service engineers to store any serializable state.   This would allow service developers to define storage entities that more closely match the needs of the business logic using application data types.

Example:

```
public class BlargEntity : OrleansTableEntity
{
    [Persistent]
    public Blarg MyBlarg; // assuming Blarg is serializable 
}
```

Supporting the use of complex types introduces the risk of developer storing complex objects as properties that, when serialized, become larger than what can fit in a single column.  To address this, I suggest a WidePersistent attribute, that indicates that the property can span multiple columns. 

Example:

```
public class BlargMapEntity : OrleansTableEntity
{
    [WidePersistent]
    public Dictionary<string,Blarg> BlargMap; // assuming Blarg is serializable 
}
```

**Advantages**
With these enhancements to OrleansAzureUtils, we can update the azure table storage provider to support entities with a column per property, and proper versioning.  It would also allow grain developers the option to use this simplified model directly, if the storage provider’s functionality is too limited.  For instance, if they want to use multiple entities or entity caching in a grain.

**Comment by jason-bragg** [comment]
As an aside, much of the code for these behaviors is being contributed by an Orleans customer, so we're not looking at developing all of this from scratch.

**Comment by veikkoeeva** [comment]
@jason-bragg
> In preparation for improving the grain state storage in Orleans, I propose the following enhancements to OrleansAzureUtils.
Is there already information available about the improvements?

**Comment by gabikliot** [comment]
From this list, the `WidePersistent` looks like the most useful for me.
Next is arbitrary table entities (especially collections, DateTime, and Guid).
`ITablePartition` seems to me personally as an over engineering - just a syntactic sugar, with potentially more confusion then benefits. But I can be convinced that I am wrong. I have less experience with building "Database Wrapping Abstractions", or ORMs. 
Just my 2 cents.

**Comment by jason-bragg** [comment]
@gabikliot 
>ITablePartition  ... syntactic sugar"

It is a simplification born of 80/20 thinking.  It's intended to reduce the cost of the common ~80% of table storage needs, while being easy to use and extensible.

> with potentially more confusion then benefits.

Potentially, but let's explore this a bit..

Cons:
- Learning curve for users already familiar with table storage.
- Limited to a simple subset of table storage operations.  Not useful for advanced scenarios.

Pros:
- Easier to learn/use for users new to azure table storage.
- Simplified usage pattern enabled common and efficient usage patterns.
- Limited abstraction layer:
  - Enables caching, monitoring, and other such extensions via decorators.
  - Improves testability of storage related application code.
  - Provides application code a degree of protection from azure storage version changes.

Are there other pros/cons I missed?

One of the biggest advantages we've found from using this pattern is that statefull grains can often be written using the simple manner one would write a stateless grain.  That is, upon receiving a grain call each call does the following: Read state, modify state, save state.  If anything goes wrong, the call simply throws.  This logic is perfectly reasonable for a stateless grain, but for a statefull one that caches the state and does not want to read from storage each time, this is too simple.

Using ITablePartition, statefull grains can use the same pattern, as the partition implementation the grain is using can be a decorator that implements a write-through cache.  Using a cached partition allows grain developers to write their grains using the simple stateless grain call pattern, but get all of the performance advantages of a statefull grain.

Further, an application can implement an IPartitionManager which decorates partitions based off of configuration settings, allowing services to modify caching/monitoring and other such behaviors at runtime using configuration switches.  Such capabilities can also be incorporated into storage providers.

**Comment by gabikliot** [comment]
I am not going to argue about that. I simply ordered the list in the priority order **in my view**. I even added and stressed: "... to me personally ... But I can be convinced that I am wrong..." No need to convince me though. If you get support of other people in the importance of that feature, go ahead.

The other two features which I ordered first - I think they are a very nice addition and will be very helpful.
